File: src/WeatherMcpServer/Tools/WeatherTools.cs
using System.ComponentModel;
using ModelContextProtocol.Server;

namespace WeatherMcpServer.Tools;

/// <summary>
/// MCP tools that provide mocked weather data.
///
/// These tools simulate a real Weather API integration.
/// In production, replace the mock data with real HTTP calls to a weather service
/// (e.g. OpenWeatherMap, National Weather Service, AccuWeather).
///
/// When Azure API Management MCP is available, these tools will be hosted behind
/// an APIM policy and the client will switch from stdio to HTTP/SSE transport.
/// </summary>
[McpServerToolType]
public sealed class WeatherTools
{
    private static readonly Dictionary<string, WeatherData> MockCurrentWeather = new(StringComparer.OrdinalIgnoreCase)
    {
        ["Seattle"] = new("Seattle, WA", 12, 54, "Cloudy with light rain", 75, 15, "SW", false),
        ["Denver"] = new("Denver, CO", 8, 46, "Partly cloudy", 40, 20, "NW", false),
        ["New York City"] = new("New York City, NY", 18, 64, "Clear", 55, 10, "NE", false),
        ["New York"] = new("New York City, NY", 18, 64, "Clear", 55, 10, "NE", false),
        ["Chicago"] = new("Chicago, IL", 5, 41, "Windy and overcast", 60, 35, "N", true),
        ["Los Angeles"] = new("Los Angeles, CA", 22, 72, "Sunny", 25, 8, "W", false),
        ["Miami"] = new("Miami, FL", 28, 82, "Thunderstorms", 85, 18, "SE", true),
    };

    private static readonly Dictionary<string, List<ForecastDay>> MockForecasts = new(StringComparer.OrdinalIgnoreCase)
    {
        ["Seattle"] = [
            new("Monday", 14, 9, "Rainy"),
            new("Tuesday", 12, 8, "Heavy rain"),
            new("Wednesday", 15, 10, "Cloudy"),
            new("Thursday", 17, 11, "Partly cloudy"),
            new("Friday", 18, 12, "Mostly sunny"),
        ],
        ["Denver"] = [
            new("Monday", 10, 2, "Sunny"),
            new("Tuesday", 12, 3, "Partly cloudy"),
            new("Wednesday", 6, -1, "Snow showers"),
            new("Thursday", 4, -3, "Heavy snow"),
            new("Friday", 8, 0, "Clearing"),
        ],
        ["New York City"] = [
            new("Monday", 20, 14, "Clear"),
            new("Tuesday", 22, 15, "Sunny"),
            new("Wednesday", 18, 13, "Partly cloudy"),
            new("Thursday", 16, 12, "Overcast"),
            new("Friday", 14, 10, "Light rain"),
        ],
        ["New York"] = [
            new("Monday", 20, 14, "Clear"),
            new("Tuesday", 22, 15, "Sunny"),
            new("Wednesday", 18, 13, "Partly cloudy"),
            new("Thursday", 16, 12, "Overcast"),
            new("Friday", 14, 10, "Light rain"),
        ],
    };

    /// <summary>
    /// Gets the current weather conditions for a given city.
    /// </summary>
    [McpServerTool(Name = "get_current_weather"), Description("Get the current weather conditions for a city.")]
    public string GetCurrentWeather(
        [Description("The city name to get weather for (e.g. 'Seattle', 'Denver', 'New York City')")]
        string city)
    {
        if (!MockCurrentWeather.TryGetValue(city, out var weather))
        {
            return $"Weather data for '{city}' is not available. " +
                   $"Supported cities: {string.Join(", ", MockCurrentWeather.Keys)}.";
        }

        return $"""
                Current weather for {weather.Location}:
                  Condition  : {weather.Condition}
                  Temperature: {weather.TempCelsius}°C / {weather.TempFahrenheit}°F
                  Humidity   : {weather.HumidityPercent}%
                  Wind       : {weather.WindSpeedKmh} km/h from the {weather.WindDirection}
                  Alerts     : {(weather.HasAlert ? "⚠️  Weather advisory in effect – check local authorities." : "None")}
                """;
    }

    /// <summary>
    /// Gets a multi-day weather forecast for a given city.
    /// </summary>
    [McpServerTool(Name = "get_weather_forecast"), Description("Get a multi-day weather forecast for a city.")]
    public string GetWeatherForecast(
        [Description("The city name to get the forecast for")]
        string city,
        [Description("Number of forecast days to return (1–5, default 5)")]
        int days = 5)
    {
        if (!MockForecasts.TryGetValue(city, out var forecast))
        {
            return $"Forecast data for '{city}' is not available. " +
                   $"Supported cities: {string.Join(", ", MockForecasts.Keys)}.";
        }

        days = Math.Clamp(days, 1, 5);
        var selected = forecast.Take(days).ToList();

        var lines = selected.Select(d =>
            $"  {d.DayName,-12} High: {d.HighCelsius}°C / {CelsiusToFahrenheit(d.HighCelsius)}°F   " +
            $"Low: {d.LowCelsius}°C / {CelsiusToFahrenheit(d.LowCelsius)}°F   {d.Condition}");

        return $"{days}-day forecast for {city}:\n" + string.Join("\n", lines);
    }

    private static int CelsiusToFahrenheit(int celsius) => celsius * 9 / 5 + 32;

    // ── Internal data models ─────────────────────────────────────────────────

    private sealed record WeatherData(
        string Location,
        int TempCelsius,
        int TempFahrenheit,
        string Condition,
        int HumidityPercent,
        int WindSpeedKmh,
        string WindDirection,
        bool HasAlert);

    private sealed record ForecastDay(
        string DayName,
        int HighCelsius,
        int LowCelsius,
        string Condition);
}

File: src/weather_mcp_server/tools/weather_tools.py
"""Mocked weather data tools used by the Weather MCP server.

These functions simulate a real Weather API integration.
In production, replace the mock data with real HTTP calls to a weather service
(e.g. OpenWeatherMap, National Weather Service, AccuWeather).

When Azure API Management MCP is available, the server hosting these tools can
be deployed as an HTTP/SSE endpoint and registered behind an APIM policy —
no code changes needed; the client switches transport via configuration.
"""

from __future__ import annotations
from dataclasses import dataclass


@dataclass(frozen=True)
class _WeatherData:
    location: str
    temp_celsius: int
    temp_fahrenheit: int
    condition: str
    humidity_percent: int
    wind_speed_kmh: int
    wind_direction: str
    has_alert: bool


@dataclass(frozen=True)
class _ForecastDay:
    day_name: str
    high_celsius: int
    low_celsius: int
    condition: str


_MOCK_CURRENT_WEATHER: dict[str, _WeatherData] = {
    k: _WeatherData(*v)
    for k, v in {
        "seattle": ("Seattle, WA", 12, 54, "Cloudy with light rain", 75, 15, "SW", False),
        "denver": ("Denver, CO", 8, 46, "Partly cloudy", 40, 20, "NW", False),
        "new york city": ("New York City, NY", 18, 64, "Clear", 55, 10, "NE", False),
        "new york": ("New York City, NY", 18, 64, "Clear", 55, 10, "NE", False),
        "chicago": ("Chicago, IL", 5, 41, "Windy and overcast", 60, 35, "N", True),
        "los angeles": ("Los Angeles, CA", 22, 72, "Sunny", 25, 8, "W", False),
        "miami": ("Miami, FL", 28, 82, "Thunderstorms", 85, 18, "SE", True),
    }.items()
}

_MOCK_FORECASTS: dict[str, list[_ForecastDay]] = {
    "seattle": [
        _ForecastDay("Monday", 14, 9, "Rainy"),
        _ForecastDay("Tuesday", 12, 8, "Heavy rain"),
        _ForecastDay("Wednesday", 15, 10, "Cloudy"),
        _ForecastDay("Thursday", 17, 11, "Partly cloudy"),
        _ForecastDay("Friday", 18, 12, "Mostly sunny"),
    ],
    "denver": [
        _ForecastDay("Monday", 10, 2, "Sunny"),
        _ForecastDay("Tuesday", 12, 3, "Partly cloudy"),
        _ForecastDay("Wednesday", 6, -1, "Snow showers"),
        _ForecastDay("Thursday", 4, -3, "Heavy snow"),
        _ForecastDay("Friday", 8, 0, "Clearing"),
    ],
    "new york city": [
        _ForecastDay("Monday", 20, 14, "Clear"),
        _ForecastDay("Tuesday", 22, 15, "Sunny"),
        _ForecastDay("Wednesday", 18, 13, "Partly cloudy"),
        _ForecastDay("Thursday", 16, 12, "Overcast"),
        _ForecastDay("Friday", 14, 10, "Light rain"),
    ],
    "new york": [
        _ForecastDay("Monday", 20, 14, "Clear"),
        _ForecastDay("Tuesday", 22, 15, "Sunny"),
        _ForecastDay("Wednesday", 18, 13, "Partly cloudy"),
        _ForecastDay("Thursday", 16, 12, "Overcast"),
        _ForecastDay("Friday", 14, 10, "Light rain"),
    ],
}


def _celsius_to_fahrenheit(celsius: int) -> int:
    return celsius * 9 // 5 + 32


def get_current_weather(city: str) -> str:
    """Get the current weather conditions for a city.

    Args:
        city: The city name (e.g. 'Seattle', 'Denver', 'New York City').

    Returns:
        A formatted string with current conditions, or an error message if
        the city is not available.
    """
    key = city.strip().lower()
    weather = _MOCK_CURRENT_WEATHER.get(key)
    if weather is None:
        supported = ", ".join(
            c.title() for c in sorted(set(_MOCK_CURRENT_WEATHER.keys()))
        )
        return (
            f"Weather data for '{city}' is not available. "
            f"Supported cities: {supported}."
        )

    alert = (
        "\u26a0\ufe0f  Weather advisory in effect \u2013 check local authorities."
        if weather.has_alert
        else "None"
    )
    return (
        f"Current weather for {weather.location}:\n"
        f"  Condition  : {weather.condition}\n"
        f"  Temperature: {weather.temp_celsius}\u00b0C / {weather.temp_fahrenheit}\u00b0F\n"
        f"  Humidity   : {weather.humidity_percent}%\n"
        f"  Wind       : {weather.wind_speed_kmh} km/h from the {weather.wind_direction}\n"
        f"  Alerts     : {alert}"
    )


def get_weather_forecast(city: str, days: int = 5) -> str:
    """Get a multi-day weather forecast for a city.

    Args:
        city: The city name to get the forecast for.
        days: Number of forecast days to return (1–5, default 5).

    Returns:
        A formatted forecast string, or an error message if the city is not
        available.
    """
    key = city.strip().lower()
    forecast = _MOCK_FORECASTS.get(key)
    if forecast is None:
        supported = ", ".join(
            c.title() for c in sorted(set(_MOCK_FORECASTS.keys()))
        )
        return (
            f"Forecast data for '{city}' is not available. "
            f"Supported cities: {supported}."
        )

    days = max(1, min(days, 5))
    selected = forecast[:days]
    lines = [
        f"  {d.day_name:<12} High: {d.high_celsius}\u00b0C / "
        f"{_celsius_to_fahrenheit(d.high_celsius)}\u00b0F   "
        f"Low: {d.low_celsius}\u00b0C / "
        f"{_celsius_to_fahrenheit(d.low_celsius)}\u00b0F   {d.condition}"
        for d in selected
    ]
    return f"{days}-day forecast for {city}:\n" + "\n".join(lines)
